### PR TITLE
Locale Suggestions: switch language based on route fragment

### LINKED ISSUE #26702: Login: Log-in page with message "Also available in English" doesn't switch language (requires refresh)
#### Steps to reproduce
1. Starting at URL: https://wordpress.com/log-in/nl?redirect_to=https%3A%2F%2Fwordpress.com%2F with browser locale set to English
2. Click "Also available in English"

#### What I expected

Page displayed in English

#### What happened instead

Page still displays Dutch, however banner is not shown. Clicking refresh shows English.

#### Browser / OS version

Chrome, Safari on macOS

#### Screenshot / Video

also available in

#### Context / Source
#dogfooding

### PR DESCRIPTION
**Pull request by ramonjd**
This PR attempts to resolve #26702 (again 😄 See: #27869)

The difference between #27869 and this PR is that we ~~call the `setUpLocale()` route middleware function _before_ `redirectDefaultLocale()`. In the latter function, we are only calling `next()` if `currentUserLocale` **[is not](https://github.com/Automattic/wp-calypso/blob/16f3c9d/client/login/controller.js#L132)** `en` but since `en` is default, `next()` is never called.~~ refresh the page after every locale change to guarantee that the language is loaded.

We now manage the dismissal state of the notice in locale storage, however the behaviour should be the same.

## What this PR does

Adds/replaces the locale fragment in the url and sets it using `window.location`. If the user dismisses the notice, we save this using `store` (locale storage).

~~Removes calls to `setLocale` in the language switcher. It's not required since the login controller takes care of language switching based on the route fragment.~~

~~Reordering the route middleware so that we switch languages first. This takes care of situations in which we switch from another language to English.~~

~~Using `page.redirect` to redirect to the login page without `en` in `redirectDefaultLocale()`. Context redirect wasn't working for non-server side route changes, for example, switching languages.~~

## Testing

1. Make sure you have English in your list of browser languages (preferably in the top 2)
2. Head to http://calypso.localhost:3000/log-in/it or http://calypso.localhost:3000/log-in/de or another non-EN language
3. Click on **Also available in English**
screen shot 2018-10-16 at 7 00 58 pm
4. Refresh the page and switch to another language

### Expectations
The languages should switch correctly to English at step 3 and to whichever language you choose at step 4.

The above instructions should also be valid for http://calypso.localhost:3000/log-in/jetpack/it and so on.

### CODE REVIEW

**Comment by alisterscott** [comment]
Thanks @ramonjd 

Hopefully this one works in production

Worked well on calypso.live except that https://hash-ecd9ef8d5de964cee45de6ef8b8e4ac4c9f0a92b.calypso.live/log-in/de showed the `/en` suffix when switching and didn't lose it on refresh as per your instructions above

**Comment by ramonjd** [comment]
> Worked well on calypso.live except that https://hash-ecd9ef8d5de964cee45de6ef8b8e4ac4c9f0a92b.calypso.live/log-in/de showed the /en suffix when switching and didn't lose it on refresh as per your instructions above

I reverted the redirect call in https://github.com/Automattic/wp-calypso/pull/28040/commits/7d85dcf1135ee3cce23c02a083b62e1cccd49fdd to use `context.redirect` again and could still reproduce this. 

https://hash-7d85dcf1135ee3cce23c02a083b62e1cccd49fdd.calypso.live/log-in/de

Looks like there is some major weirdness going on.

**Comment by alisterscott** [comment]
@ramonjd is there anything we can do with this one?

**Comment by ramonjd** [comment]
@alisterscott This one is bugging me, since we can't test it reliably before production, which makes me think we shouldn't merge it. It's a shame because the bug is visible enough to be annoying. 

The only fail-proof alternative I can think of right now is that we do a hard refresh of the page after a language switch. That'll work.

**Comment by ramonjd** [comment]
> Worked well on calypso.live except that https://hash-ecd9ef8d5de964cee45de6ef8b8e4ac4c9f0a92b.calypso.live/log-in/de showed the /en suffix when switching and didn't lose it on refresh as per your instructions above

@alisterscott  After implementing and testing the hard refresh version, I realised it wasn't related to ⬆️ ... so I reverted that and did some more debugging, and found that the `{hash}.live` links, for some reason don't drop the `en` anyway. I checked a few other branches to confirm this. [Example](https://hash-7bd014ee80dc1f680bec5f5a758e7585e5273cf1.calypso.live/log-in/en).

I suspect they're not passing [this check](https://github.com/Automattic/wp-calypso/blob/79d5278/client/login/controller.js#L124), though I don't know why. 

Long story short, I think that if we've fixed the bug that causes English to not load properly, there's a good chance that the route fragment issue might resolve itself (though we can only test in staging :()

**Comment by alisterscott** [comment]
Thanks @ramonjd 

The canary failure is because this needs a fresh rebase from master to get the new gutenberg routes

**Comment by ramonjd** [comment]
Closing this. It's a bit stale. Will revisit.
